src/sage/rings/polynomial/integer_valued_polynomials.py

```diff
@@ -0,0 +1,1106 @@
+                sage: F._realization_name()
+                'shifted'
+            """
+            return "shifted"
+
+        def product_on_basis(self, n1, n2):
+            r"""
+            Return the product of basis elements ``n1`` and ``n2``.
+
+            INPUT:
+
+            - ``n1``, ``n2`` -- integers
+
+            EXAMPLES::
+
+                sage: A = IntegerValuedPolynomialRing(QQ).S()
+                sage: A.product_on_basis(0, 1)
+                S[1]
+                sage: A.product_on_basis(1, 2)
+                -2*S[2] + 3*S[3]
+            """
+            i = ZZ(n1)
+            j = ZZ(n2)
+            if j < i:
+                j, i = i, j
+
+            R = self.base_ring()
+            return self._from_dict({i + j - k: R((-1)**k * i.binomial(k) * (i + j - k).binomial(i))
+                                    for k in range(i + 1)})
+
+        def _from_binomial_basis(self, i):
+            """
+            Convert from the ``binomial(x,k)`` basis.
+
+            INPUT:
+
+            - ``i`` -- an integer
+
+            EXAMPLES::
+
+                sage: S = IntegerValuedPolynomialRing(ZZ).S()
+                sage: B = IntegerValuedPolynomialRing(ZZ).B()
+                sage: b = B.basis()
+                sage: S(b[3]+1)  # indirect doctest
+                3*S[1] - 3*S[2] + S[3]
+                sage: B(_)
+                B[0] + B[3]
+            """
+            i = ZZ(i)
+            R = self.base_ring()
+            return self._from_dict({k: R((-1)**(i - k) * i.binomial(k))
+                                    for k in range(i + 1)})
+
+        def from_h_vector(self, h):
+            """
+            Convert from some `h`-vector.
+
+            INPUT:
+
+            - ``h`` -- a tuple or vector
+
+            .. SEEALSO:: :meth:`Element.h_vector`
+
+            EXAMPLES::
+
+                sage: A = IntegerValuedPolynomialRing(ZZ).S()
+                sage: S = A.basis()
+                sage: ex = S[2]+S[4]
+                sage: A.from_h_vector(ex.h_vector())
+                S[2] + S[4]
+            """
+            d = len(h) - 1
+            m = matrix(QQ, d + 1, d + 1,
+                       lambda j, i: (-1)**(d - j) * binomial(d - i, d - j))
+            v = vector(QQ, [h[i] for i in range(d + 1)])
+            return self._from_dict({i: Integer(c)
+                                    for i, c in enumerate(m * v)})
+
+        def _element_constructor_(self, x):
+            r"""
+            Convert ``x`` into ``self``.
+
+            INPUT:
+
+            - ``x`` -- an integer or something convertible
+
+            EXAMPLES::
+
+                sage: R = IntegerValuedPolynomialRing(QQ).S()
+                sage: x = R.gen()
+                sage: R(3)
+                3*S[0]
+                sage: R(x)
+                S[1]
+            """
+            if x in NonNegativeIntegers():
+                W = self.basis().keys()
+                return self.monomial(W(x))
```

Comment:
I am not sure about this. You get very different behaviors for `-2` and `2`. I feel like the output type should generally depend on the parent. However, I guess this should never be called whenever `ZZ` is a subring of the base ring as the coercion should get it first as per the doctest. So perhaps over a positive characteristic field this would be invoked...but it seems sketchy to me.